### PR TITLE
Jspada/signer and hasher extensions

### LINKED ISSUE #418: Extend signer crate into a hasher and signer

### PR DESCRIPTION
**Pull request by jspada**
Closes #418

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 [workspace]
 members = [
     "cairo",
+    "crypto",
```
Comment (querolita) [inline Cargo.toml:4]: Actually I don't know if `crypto` could be a misleading/too general name...

File: crypto/README.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+* The [signer](crate::signer) follows the algorithm outlined in the [Mina Signature Specification](https://github.com/MinaProtocol/mina/blob/master/docs/specs/signatures/description.md)
+
+## Hasher interface
+
+The `hasher` module currently supports creating both legacy and an experimental kimchi hashers.
+
+* [`hasher::create_legacy`] create a legacy hasher
+* [`hasher::create_kimchi`] create an experimental kimchi hasher
+
+Here is an example of how to use the hasher interface.
```
Comment (querolita) [inline crypto/README.md:15]: It's so nice you're giving examples as part of the readme

File: crypto/src/hasher/mod.rs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+///
+/// See example in [`ROInput`] documentation
+pub trait Hashable: Clone {
+    /// Generic domain string argument type
+    type D: DomainParameter;
+
+    /// Serialization to random oracle input
+    fn to_roinput(self) -> ROInput;
+
+    /// Generate unique domain string of length `<= 20`.
```
Comment (querolita) [inline crypto/src/hasher/mod.rs:113]: perhaps clarify in the doc that the uniqueness and length bound are not imposed for free, but should be enforced by the developer implementing the trait

**Review by mimoo — COMMENTED** [review]
LGTM, except for the naming of crypto/ which is a bit too generic

File: crypto/src/hasher/poseidon.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            phantom: PhantomData,
+        };
+
+        poseidon.init(domain_param);
+
+        poseidon
+    }
+}
+
+pub(crate) fn new_legacy<H: Hashable>(domain_param: H::D) -> impl Hasher<H> {
```
Comment (querolita) [inline crypto/src/hasher/poseidon.rs:47]: We don't impose a particular return type here as opposed to `new_kimchi` ?

**Review by querolita — APPROVED** [review]
Very elegant solution

File: crypto/src/hasher/mod.rs (hunk 3 above)
Comment (mimoo) [reply crypto/src/hasher/mod.rs:113]: can you assert this somewhere though?

File: crypto/README.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
+# Mina crypto
+
+This crate provides an API and framework for Mina hashing and signing.
+
+* The [hasher](crate::hasher) is a safe wrapper around Mina's instances of the [Poseidon arithmetic sponge](https://github.com/o1-labs/cryptography-rfcs/blob/master/mina/001-poseidon-sponge.md)
+* The [signer](crate::signer) follows the algorithm outlined in the [Mina Signature Specification](https://github.com/MinaProtocol/mina/blob/master/docs/specs/signatures/description.md)
+
+## Hasher interface
+
+The `hasher` module currently supports creating both legacy and an experimental kimchi hashers.
```
[the remark this answers is not in the document]
Comment (jspada) [reply crypto/README.md:10]: It's meant to :-) The new ROInput format is not yet implemented for it.

File: crypto/README.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+use mina_crypto::hasher::{create_legacy, Hashable, Hasher, ROInput};
+
+#[derive(Clone)]
+struct Example {
+    x: u32,
+    y: u64,
+}
+
+impl Hashable for Example {
+    type D = u32;
```
[the remark this answers is not in the document]
Comment (jspada) [reply crypto/README.md:27]: It's the type of parameter from which the domain separation string is generated.  For example, for mina transactions it's a NetworkId (testnet or mainnet) and for Merkle nodes it's the merkle tree height (u64).  In mina different hashes use different types of parameters to seed the domain strings.

File: crypto/src/hasher/mod.rs (hunk 3 above)
Comment (jspada) [reply crypto/src/hasher/mod.rs:113]: Yes and it already is. 👍🏼

Comment (jspada) [reply crypto/src/hasher/mod.rs:113]: There's already a comment on the trait warning that the uniqueness bound must be assured by the developer, but I will also add a note about this on the method itself.  Good suggestion - thanks.

File: crypto/src/hasher/poseidon.rs (hunk 4 above)
Comment (jspada) [reply crypto/src/hasher/poseidon.rs:47]: Thanks-- I'll change to be consistent.  👍🏼

File: crypto/README.md [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            );
+
+let mut ctx = mina_crypto::signer::create_legacy::<Transaction>(NetworkId::TESTNET);
+let sig = ctx.sign(keypair, tx);
+assert!(ctx.verify(sig, keypair.public,tx));
+```
+
+Note that these examples use the test [`Transaction`](https://github.com/o1-labs/proof-systems/tree/master/crypto/tests/transaction.rs) structure found in the [`./tests`](https://github.com/o1-labs/proof-systems/tree/master/crypto/tests) directory.  This is a complete reference implementation of the Mina payment and delegation transaction structures found on mainnet and testnet.
+
+## Framework
```
[the remark this answers is not in the document]
Comment (jspada) [reply crypto/README.md:101]: Fixed!

File: Cargo.toml (hunk 1 above)
Comment (jspada) [reply Cargo.toml:4]: I'll split it up as David suggested.

File: crypto/Cargo.toml [hunk 8]
```diff
@@ -24,3 +24,5 @@ bitvec = "1.0.0"
 sha2 = "0.10.0"
 bs58 = "0.4.0"
 thiserror = "1.0.30"
+serde = { version = "1.0", features = ["derive"] }
+serde_json = { version = "1.0" }
```
Comment (mimoo) [inline crypto/Cargo.toml]: missing EOL

File: crypto/README.md [hunk 9]
```diff
@@ -167,10 +169,10 @@ impl Hashable for ExampleMerkleNode {
         roi
     }
 
-    fn domain_string(this: Option<Self>, _: &Self::D) -> String {
+    fn domain_string(this: Option<Self>, _: &Self::D) -> Option<String> {
         match this {
             None => panic!("missing this argument (should never happen)"),
-            Some(x) => format!("ExampleMerkleNode{:03}", x.height),
+            Some(x) => format!("ExampleMerkleNode{:03}", x.height).into(),
         }
```
Comment (mimoo) [inline crypto/README.md]: btw you could write these 4 lines as 

```rust
this.map(|x| format!("ExampleMerkleNode{:03}", x.height)).ok_or_else(|| panic!("missing this argument (should never happen)")
```
```

File: crypto/README.md [hunk 10]
```diff
@@ -270,14 +272,16 @@ impl Hashable for B {
...
+        // Way 2: Append ROInput
         roi.append_roinput(self.a2.to_roinput());
         roi.append_u32(self.z);
         roi
     }
 
-    fn domain_string(_: Option<Self>, _: &Self::D) -> String {
-        format!("B")
+    fn domain_string(_: Option<Self>, _: &Self::D) -> Option<String> {
+        format!("B").into()
```
Comment (mimoo) [inline crypto/README.md]: btw I think `Some(format!("B"))` is better

File: hasher/Cargo.toml [hunk 11]
```diff
@@ -12,19 +12,8 @@ oracle = { path = "../oracle" }
...
 bitvec = "1.0.0"
-sha2 = "0.10.0"
-bs58 = "0.4.0"
-thiserror = "1.0.30"
 serde = { version = "1.0", features = ["derive"] }
-serde_json = { version = "1.0" }
-
-[dev-dependencies]
-array-init = "2.0.0"
+serde_json = { version = "1.0" }
```
Comment (mimoo) [inline hasher/Cargo.toml:19]: EOL missing

**Review by mimoo — APPROVED** [review]
looks good

**Pull request merged**